File: programmers/js/stack-que/기능개발.js
/*
    문제: 기능개발
    출처: 프로그래머스 Level 2
    유형: 스택
    핵심 아이디어: 작업량 더해가며 가장 앞/뒤부터 배포 가능성 확인
    배운점:
        - JS LIFO
    시간복잡도: TODO
*/

// pop 안쓰기
function solution(progresses, speeds) {
    var answer = [];
    
    while (progresses.length > 0) {
        progresses = progresses.map((p, i) => p + speeds[i]);
        let count = 0;
        
        while (progresses[0] >= 100) {
            count ++;
        }
        
//         배포 가능한 프로그램 수만큼 앞에서부터 슬라이스
        if (count > 0) {
            progresses.slice(0, count);
            answer.push(count);
        };
    }
    
    return answer;
}

// reverse 해서 pop 쓴 것 JS stack 이라서 pop 씀 (LIFO)
function solution(progresses, speeds) {
    var answer = [];
    
    progresses.reverse();
    speeds.reverse();
    
    while (progresses.length > 0) {
        // 진행도 더해주고, 프로그램 수 초기화
        progresses = progresses.map((p, i) => p + speeds[i]);
        let count = 0;
        
        // 진행도가 채워지면 pop, 완료한 프로그램만큼 세주기
        while (progresses[progresses.length-1] >= 100) {
            progresses.pop();
            count ++;
        }
        // 정답 배열에 완료한 프로그램 수 넣어주기
        if (count > 0) answer.push(count);
    }
    
    return answer;
}

File: programmers/py/stack_que/기능개발.py
"""
    문제: 기능개발
    출처: 프로그래머스 Level 2
    유형: 큐 / Grouping
    핵심 아이디어: deque써서 시뮬레이션 or 한번에 며칠 걸리는지 계산
    배운점: -
    시간복잡도:
        - deque: O(n^2)
        - 날짜 계산: O(n)
"""
from collections import deque

# deque 사용
def solution(progresses, speeds):
    p = deque(progresses)
    s = deque(speeds)
    answer = []
    
    while p:
        for i in range(len(p)):
            p[i] += s[i]
            
        c = 0
        while p and p[0] >= 100:
            p.popleft()
            s.popleft()
            c += 1
            
        if c > 0:
            answer.append(c)

    return answer

def solution(progresses, speeds):
#     날짜 계산
    days = []
    for p, s in zip(progresses, speeds):
        days.append(math.ceil((100 - p)/s))
    
#     기능 개발 완료되는 시점 묶기
    answer = []
    current = days[0]
    count = 1
    
    for i in range(1, len(days)):
        if days[i] <= current:
            count += 1
        else:
            answer.append(count)
            current = days[i]
            count = 1
    answer.append(count)
    
    return answer